tests/test_dataset.py

```diff
@@ -0,0 +1,204 @@
+        self.assertObject(result, dataset)
+
+
+    def test_dataset_get_404(self):
+        with self.assertRaises(pyrefinebio.exceptions.NotFound):
+            pyrefinebio.Dataset.get("this-does-not-exist")
+
+    
+    @patch("pyrefinebio.http.requests.request", side_effect=mock_request)
+    def test_dataset_500(self, mock_request):
+        with self.assertRaises(pyrefinebio.exceptions.ServerError):
+            pyrefinebio.Dataset.get("500")
+
+
+    @patch("pyrefinebio.dataset.post_by_endpoint")
+    @patch("pyrefinebio.dataset.put_by_endpoint")
+    def test_dataset_save(self, mock_put, mock_post):
+        mock_post.return_value = dataset
+
+        ds = pyrefinebio.Dataset(data={"test-experiment": ["sample-1", "sample-2"]}, email_address="test-email")
+        ds = ds.save()
+
+        mock_post.assert_called_with(
+            "dataset",
+            payload={
+                "data": {"test-experiment": ["sample-1", "sample-2"]},
+                "email_address": "test-email"
+            }
+        )
+    
+        ds.email_address = "changed-email"
+        ds.save()
+
+        mock_put.assert_called_with(
+            "dataset/test-dataset",
+            payload={
+                "data": {"test-experiment": ["sample-1", "sample-2"]},
+                "aggregate_by": "EXPERIMENT",
+                'scale_by': 'NONE',
+                'email_address': 'changed-email',
+                'quantile_normalize': True,
+                'svd_algorithm': 'NONE'
+            }
+        )
+
+
+    def test_dataset_save_bad_data(self):
+        with self.assertRaises(pyrefinebio.exceptions.BadRequest):
+            pyrefinebio.Dataset(data={"lol": ["not-good"]}).save()
+
+
+    @patch("pyrefinebio.http.requests.request", side_effect=mock_request)
+    def test_dataset_process(self, mock_request):
+        ds = pyrefinebio.Dataset(data={"test-experiment": ["sample-1", "sample-2"]}, email_address="test-email")
+        ds.process()
+
+        self.assertTrue(ds.start)
+        self.assertTrue(ds.is_processing)
+
+
+    @patch("pyrefinebio.config.os.path.exists")
+    @patch("pyrefinebio.config.yaml.full_load")
+    @patch("pyrefinebio.config.open")
+    def test_dataset_process_bad_token(self, mock_open, mock_load, mock_exists):
+        Config._instance = None
+        os.environ["CONFIG_FILE"] = "./temp"
+        mock_open.return_value.__enter__.return_value = "file"
+        mock_load.return_value = {"token": "this-is-a-bad-token"}
+        mock_exists.return_value = True
+
+        data = {}
+        exs = pyrefinebio.Experiment.search(num_downloadable_samples=1)
+        ex = next(exs)
+        data[ex.accession_code] = ["ALL"]
+
+        ds = pyrefinebio.Dataset(data=data)
+
+        with self.assertRaises(pyrefinebio.exceptions.BadRequest) as br:
+            ds.process()
+        
+        e = br.exception
+
+        self.assertEqual(br.exception.base_message, "Bad Request: ['You must provide an active API token ID']")
```

Comment:
Oh I see we do have a message specific to the token. I feel like we might as well make it it's own exception type for clarity.

Reply:
You know what, I think this is clear enough.

Reply:
I think as a part of AlexsLemonade/refinebio#2546 I might try to make all errors returned by the api as uniformed as possible. I was thinking that they should all look like:
```json
{
    "error-type": "bad-token",
    "message": "You must provide an active API token ID",
    "details": "pass an active token id as a header with key `API-KEY`"
}
```
I think it would make the api more user friendly and it would make handling errors in the client a lot easier. We could just make a custom exception for every `error-type`. lmk what you think though.